### ISSUE TITLE
Destructive behaviour of `set_array`

### ISSUE THREAD

**Issue by TobiasNx** [metafacture/metafacture-fix#111]
Since we do not support the creation of new arrays with `$append`/`$prepend` because of #110 we are depending on `set_array` to create an array before adding elements or objects to that array.

But if the array could exist but does not always exist we have a problem with `set_array` as it is destructive.
See:
https://github.com/TobiasNx/fix-FunctionalReview-Testing/tree/master/data/testing/set_array_existing_array

We should reintroduce $append and $prepend as creation methods if the array does not exist in a record to add to or create an array.

**Comment by TobiasNx** [comment]
Additional problem with the need of set_array it creates empty arrays if the values that should be added do not exist in an record.
```
{
a: b
b: d
}
{
a:b
}
```

>>
```
set_array("e[]")
move_field("b", "e[].4append")
```

Results in:
```
{
a: b
e: [d]
}
{
a:b
e: [ ]
}
```

**Comment by TobiasNx** [comment]
Additional problem(?) is the inconsistent handling the appending without set_array, if there is only one field that is appended it creates an JSON-Array even without set_array, if there are more it does not:

https://github.com/TobiasNx/fix-FunctionalReview-Testing/tree/master/data/testing/singleAppending

Without additional set_array, it should:
```
{
  "animal": "bunny",
  "stringimals": [
    {
      "1": "bunny"
    }
  ],
  "strongimals": [
    "bunny"
  ]
}
{
  "animal": "bunny",
  "bnimal": "bunny",
  "stringimals": [
    {
      "1": "bunny",
      "2": "bunny"
    }
  ],
  "strongimals": [
    "bunny",
    "bunny"
  ]
}
```
But it outputs:
```
{
  "animal" : "bunny",
  "stringimals" : [ "bunny" ],
  "strongimals" : [ "bunny" ]
}
{
  "animal" : "bunny",
  "bnimal" : "bunny",
  "stringimals" : [ {
    "1" : "bunny",
    "2" : "bunny"
  } ],
  "strongimals" : [ "bunny", "bunny" ]
}
```

**Comment by blackwinter** [comment]
Catmandu preserves existing contents for both [arrays](https://github.com/hbz/Catmandu/blob/dev/t/Catmandu-Fix-set_array.t#L28) and [hashes](https://github.com/hbz/Catmandu/blob/dev/t/Catmandu-Fix-set_hash.t#L28).

**Comment by fsteeg** [comment]
Will revisit after #102.

**Comment by blackwinter** [comment]
Catmandu is destructive! See tests for [`set_array`](https://github.com/hbz/Catmandu/blob/metafix-compatibility/t/Catmandu-Fix-set_array.t#L31-L38) and [`set_hash`](https://github.com/hbz/Catmandu/blob/metafix-compatibility/t/Catmandu-Fix-set_hash.t#L31-L38).

**Comment by TobiasNx** [comment]
This can be closed, if catmandu is desctructive too? @blackwinter

**Comment by blackwinter** [comment]
As soon as the current state of `102-path` (1a3278a) is merged.

**Comment by fsteeg** [comment]
> This can be closed [...] As soon as [...] (https://github.com/metafacture/metafacture-fix/commit/1a3278a92f9fc020fb84f88afe31c6d2ff6a107e) is merged.

1a3278a92f9fc020fb84f88afe31c6d2ff6a107e was merged with 15e223749e3130bf5adac4961fa2954b6a7b4f40, closing.
